cmd/controller/run.go

```diff
@@ -34,8 +34,12 @@ import (
 )
 
 const (
-	PprofListenAddr       = "0.0.0.0:6060"
-	kappctrlAPIPORTEnvKey = "KAPPCTRL_API_PORT"
+	PprofListenAddr                  = "0.0.0.0:6060"
+	kappctrlAPIPORTEnvKey            = "KAPPCTRL_API_PORT"
+	kappctrlReconcileNamespaceEnvKey = "RECONCILE_NAMESPCE"
+	kappctrlReconcileScopeEnvKey     = "RECONCILE_SCOPE"
+	cluster                          = "cluster"
+	namespace                        = "namespace"
```

Comment:
per the comment I left earlier here is a slightly more fleshed out thought:
Here, rather than the "ReconcileScope" flag, we could introduce a boolean flag for "should i run the aggregate api server?" (yes/no) and an orthogonal environment variable for "what is the scope of the namespaces i'm reconciling?"    this second env var could use empty string ("") to indicate "all namespaces" or a legal namespace name and do an exact string match.  that way we could avoid regular expressions. :) 

so then the env vars would become something more like 
```
KAPPCTRL_START_API_SERVER
KAPPCTRL_RECONCILE_NAMESPACE
```

further, as we do prefer flags, and it turns out there's already a namespace flag: 
https://github.com/vmware-tanzu/carvel-kapp-controller/blob/8dba0472c9ebdca83d3dd981f783ff9afa75ac2b/cmd/controller/main.go#L24

then it's only necessary to add a boolean flag indicating whether or not to start the aggregate api server. 

Note that in this arrangement we would also be namespacing the PKGR resource - this was called out in a different comment but maybe now it's more clear why that's preferred, since we're slicing the functionality slightly differently.

Reply:
If we use KAPPCTRL_START_API_SERVER and KAPPCTRL_RECONCILE_NAMESPACE, it means we need to add another constant for namespace, see the sheet below:
![image](https://user-images.githubusercontent.com/1787505/181705230-e6325c68-4e13-4db1-abe9-92e575def7d3.png)
Is it right?

For my point of view, reconcile scope and start api server just have indirect relationship. Is reconcile scope more straightforward to user?

Reply:
Hi @showpune - I don't think the above diagram makes sense, no.  I don't think we need that middle row, because I think all the reconcilers should be scoped. By scoping all the reconcilers together: app, pkgi, pkgr, pkg; all of those either are directed towards one namespace or all namespaces,  then the only question is "should this process also start the apiserver?"   

We definitely want to avoid magic strings, as for example in your example above a namespace called "None" would create problems.

If you're concerned that the process running the aggregate api server shouldn't also be reconciling anything at all, one option could be to point it at a namespace where we don't expect any resources to be created (such as the `kapp-controller` namespace of kapp-controller's deployment). 

Another option is to use regular expressions to specify which namespaces to match. This would introduce regular expressions with all the problems they bring, but it could also allow for an administrator to avoid needing to add an additional kapp-controller for every single namespace. (for example, maybe you'd want a single kapp-controller to reconcile all namespaces matching `dev-.*` to simplify developer onboarding?) 

I agree that reconcile scope and whether or not start the API server are unrelated. I think that's a point in favor of slicing the parameters this way, as it avoids having dependencies across flags.  As far as names of the flags, I think you would have better "user perspective" than I do, but we can also ask in our slack channel or community meeting.

does this make more sense, what I'm trying to propose?  What do you think?

Reply:
Hi @joe-kimmel-vmw，

It is OK that we use KAPPCTRL_START_API_SERVER and KAPPCTRL_RECONCILE_NAMESPACE. 

I want to confirm with you that if KAPPCTRL_START_API_SERVER is set to false, should we start reconciler for cluster resource like pkgr and pkg? The deployments are like below:
![image](https://user-images.githubusercontent.com/1787505/183335281-ccb01917-3ca2-499b-bbf4-45e9acd96a3f.png)
There should be a controller in each namespace, if we start reconciler in each namespace, if anything changed in cluster resource, it will be reconciled by all namespace with reconciler and there should be some data conflict issue

So the solution should be:
KAPPCTRL_START_API_SERVER=true: start the api server, reconcile the cluster resource like pkg and pkgr
KAPPCTRL_START_API_SERVER=false: not start the api server, not reconcile the cluster resource like pkg and pkgr

Is it right?

Reply:
@showpune I think there's still a misunderstanding. The way i see it, _all_ resources that are reconciled are namespaced. I'm not sure that we need the notion of a "cluster resource".  Am I missing something?

Reply:
Hi @joe-kimmel-vmw ,

I understood your idea and I commit the change, is it what you expect?